### ISSUE TITLE
getCases method requires suite id and fails when attempting to run.

### ISSUE THREAD

**Issue by voodoofish** [dlenroc/node-testrail-api#46]
This was run against a setup using 7.0.1

A second issue  I noted after updating the method locally to work was that in some larger sets(600-3000), The method would only get a subset. 341 of 600+ or 900 of 2000. 
I left the default limit as adding the limit would just get that amount vs not setting the limit would get 250+ records when there were more.
I tried debugging the issue and noted(in the pagination "while" loop), that the first call would work, limit = 250, offset = 0, but on the second call, with the offset = 250, I would end up just getting 140 vs the 250+ remaining tests.
I would also see, if I console logged the output of the getCases, a line at the end that would say "... xxx more items"
however,  if I do the calls serially with the limit and offset passed in, I would get the correct values.
This seems like what the while loop is doing so I'm not clear as to why it's failing. 
ex
getcases(projId, suiteId, {limit: 250, offset: 0) // returns 250
getcases(projId, suiteId, {limit: 250, offset: 250) // returns 250
getcases(projId, suiteId, {limit: 250, offset: 500)// returns 25

EDIT: A little more debugging and seems like it's at the item filter. It catching duplicates that appear in the  second set of data.
Interestingly enough, if I remove the line to not filter, it gets the expected number of cases...
Now I'm wondering if it's an api issue.

EDIT2: Not an api error. Further looking, it looks like there is a var row_number that exists that can help in uniqeifying the Set value used to compare.
Both changes are in my forked repo.

**Comment by dlenroc** [comment]
Thanks @voodoofish you've made an awesome work investigating the case and providing a solution

1. about `getCases method requires suite id and fails when attempting to run`
    
    suite_id is sent as a query parameter, not as a path parameter that is why it won't have a separate method

    so the right usage for projects that require sending suite_id is ⬇️
    
    ```typescript
     api.getCases(projectId, { suite_id: suiteId })
    ```
   
     > Editor/IDE must suggest them too: https://github.com/dlenroc/node-testrail-api/blob/main/src/payload/request/CaseFilters.ts

2. about `pagination`

    Before going to the problems let me describe a bit how pagination in that project is going to work in order to be on the same page (will update README.md through a separate PR later too)

    Build-in pagination must work only in cases when `limit` and `offset` are not explicitly set (that allow implementing own concurrent pagination and at the same time preserve compatibility between multiple versions of the TestRail - in the new versions some endpoints have pagination which wasn't in previous)

    > case 1 - auto pagination is disabled when limit or offset is specified
    ```typescript
    api.getCase(projectId, { limit: 10 }) // fist 10 items must be returned
    api.getCase(projectId, { offset: 10 }) // 10 skiped and up to 250 items are returned (testrail backend default)
    api.getCase(projectId, { limit: 10, offset: 10 }) // second 10 items must be returned

    // note: suite_id and other CaseFilters options can be used too /src/payload/request/CaseFilters.ts
    ```

     > case 2 - auto pagination is enabled, 250 items per page is used (max limit allowed by TestRail)
    ```typescript
    api.getCase(projectId) // all cases for regular project (withoult suites)
    api.getCase(projectId, { suite_id: suiteId }) // all cases from specific suite

    // note: suite_id and other CaseFilters options can be used too /src/payload/request/CaseFilters.ts
    ```

    I looked at your solution, and yes filter must be updated, but I don't see `row_number` field either in TestRail documents or the TestRail instances I use, soo can you please check and confirm if you really have different cases with the same ID and where `row_number` come from (having duplicate IDs looks like a bug or another TestRail bug/feature I'm not aware of)

    Also can you please check if the issue is reproducible with `@dlenroc/testrail@1.6.1`

   ```shell
   npm install @dlenroc/testrail@1.6.1
   ```

**Comment by voodoofish** [comment]
Thanks for looking into this and the clarification. I'm fairly new to typescript/JS in general so this was a fun little exercise in trying to figure out why things weren't working.

Comment # 1
Totally makes sense to pass as a param and I understand why. Again, thanks for the clarification.

Comment # 2
Great explanation of the pagination. I think I largely understood it as you noted but this def fills in some of the gaps.
So row_number appeared in what appeared to be duplicate records(as caught by the initial filter, just the Id) as well as when both id and section were identical in the first attempt at fine tuning the filter . I was curious about the value as well as I noted it wasn't part of the documented response object which makes me wonder if it was just not updated in the docs. My initial conclusion was based on the way that a case/section could be copied and it being an additional attribute for those copies. This seemed to match the data I was looking at for at least one of the sets of cases.
There was always a  case that had no row_number and an identical one with a row number. What the row_number signifies is a question though as it would need to be tied to another section somehow....
Since the number of cases that show up in the ui matches the number of cases I get back with the filter I had set up I figured this was a good possibility. 

I also had a data set that had the row_number and no matching id with other records. This could be attributed to it being copied from a different suite. I would need to investigate further though as it was at the end of the day and I was just happy to have all of the project case data...  :)

I can try the latest version(1.6.1) on Monday and let you know what I find.

Cheers!

**Comment by voodoofish** [comment]
Hi @dlenroc ,

I posted a question on the gurock forums in relation to the row_number to see if I can get some info.
[This is the link](https://discuss.gurock.com/t/api-question-about-result-values-of-get-cases/21284)  to check it out/follow/add to if you would like.

So I ran the updated version filter against my data and I do get a higher result count, but it still filters out some records.
As for the results when I don't filter all of the values and in the sample project/suite I'm looking at, I do have 2 records that share id and section_id and with those two records, one contains the row_number value while the other does not return a row value.
I also attempted to edit the test name to see if the result set would be different.
So the results were interesting. In doing a text search, I only found the single record I had edited. when looking at the unfiltered output I had 2 records with the identical string. Though again, one had a row_number value while the other didn't....so...scratching my head on this one and waiting for the response from Gurock on this.
